File: firstDuplicate.js
//How can an algorithm break up the code into some like this? 
// # def firstDuplicate(a):
// #     for i in a:
// #         a[abs(i) - 1] *= -1
// #         if a[abs(i) - 1] > 0:
// #             return abs(i)
// #     return -1

// INTO:
// function firstDuplicate(array = []) {
//     function iterate(array) {

//     }

//     function multiply(array, idx) {

//     }

//     function greaterThanZero(array, idx) {

//     }

//     function absolute(num) {

//     }

// Passes codesignal.
// I think this abstraction is sufficient for a student although it could be abstracted more.
function firstDuplicate(array = []) {
    function iterate(array) {
        for (n of array) {
            multiply(array, n);
            if (greaterThanZero(array, n)) return n;
        }
    }

    function multiply(array, idx) {
        return array[absolute(idx) - 1] *= -1;
    }

    function greaterThanZero(array, idx) {
        return array[absolute(idx) -1 ] > 0;
    }

    function absolute(num) {
        return Math.absolute(num);
    }

    result = iterate(array);
    return result ? result : -1;
}

File: FirstDuplicate.java
import java.lang.Math;
public class FirstDuplicate {
    public static void main(String[] args) {
        // Integer[] array = { 8,4,6,2,6,4,7,9,5,8 };
        Integer[] array = { 5, 5, 5, 5, 5 };
        Integer result = iterate(array);
        System.out.println(result != null ? result : -1);
    }

    public static Integer iterate(Integer[] array) {
        for (Integer n : array) {
            multiply(array, n);
            if (greaterThanZero(array, n)) return n;
        }
        return 0;
    }

    public static void multiply(Integer[] array, Integer idx) {
        array[absolute(idx) - 1] *= -1;
    }

    public static Boolean greaterThanZero(Integer[] array, Integer idx) {
        return array[absolute(idx) - 1] > 0;
    }

    public static Integer absolute(Integer num) {
        return Math.abs(num);
    }
}
